src/compiler/checker.ts

```diff
@@ -1899,6 +1902,23 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         typeHasCallOrConstructSignatures,
     };
 
+    function getContextualStringLiteralCompletionTypes(expression: Expression) {
+        const seen = new Map<string, true>();
+
+        return [
+            ...getStringLiteralTypes(getContextualTypeFromParent(expression, checker, ContextFlags.None), seen),
+            ...getStringLiteralTypes(getContextualTypeFromParent(expression, checker, ContextFlags.Completions), seen),
+        ];
+    }
+
+    function getStringLiteralTypes(type: Type | undefined, uniques = new Map<string, true>()): readonly StringLiteralType[] {

```

Comment:
I'd prefer either a more specific name or pushing the function inside `getContextualStringLiteralCompletionTypes`